cmd/scCount/scCount.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/gtf"
+	"github.com/vertgenlab/gonomics/interval"
+	"github.com/vertgenlab/gonomics/sam"
+	"io"
+	"log"
+	"sort"
+	"strings"
+)
+
+func scCount(s Settings) {
+	var sc sam.SingleCellAlignment
+	var err error
+	readChan, _ := sam.GoReadToChan(s.InFile)
+	genes := gtf.Read(s.GeneFile)
+	var geneIntervals []interval.Interval
+	var geneIdSlice []string = make([]string, 0)
+	var geneIndex = make(map[string]int)
+	var headerString string = "Bx"
+
+	for g := range genes {
+		geneIdSlice = append(geneIdSlice, g)
+	}
+	sort.Strings(geneIdSlice)
+	for c, g := range geneIdSlice { //c for count, g for gene
+		headerString += fmt.Sprintf("\t%s", g)
+		geneIndex[genes[g].GeneID] = c
+		geneIntervals = append(geneIntervals, genes[g])
+		c++
+	}
+
+	out := fileio.EasyCreate(s.OutFile)
+	_, err = fmt.Fprintf(out, "%s\n", headerString)
+	exception.PanicOnErr(err)
+	tree := interval.BuildTree(geneIntervals)
+	var overlap []interval.Interval
+	var currRow Row
+	var currGene string
+	var firstTime bool = true
+	var normFlag bool = false
+
+	var normalizationMap map[string]float64 = make(map[string]float64, 0)
+	if s.ExpNormalizationFile != "" {
+		normFlag = true
+		var line string
+		var words []string
+		var doneReading bool = false
+		exp := fileio.EasyOpen(s.ExpNormalizationFile)
+		for line, doneReading = fileio.EasyNextRealLine(exp); !doneReading; line, doneReading = fileio.EasyNextRealLine(exp) {
+			words = strings.Split(line, "\t")
+			if len(words) != 2 {
+				log.Fatalf("Expression normalization input file must be a tab-separated file with two columns per line.")
+			}
+			normalizationMap[words[0]] = common.StringToFloat64(words[1])
+		}
+		err = exp.Close()
+		exception.PanicOnErr(err)
+	}
+
+	for i := range readChan {
+		sc = sam.ToSingleCellAlignment(i)
+		overlap = interval.Query(tree, &i, "any")
+		if len(overlap) == 0 {
+			continue
+		}
+		if len(overlap) > 1 {
+			//TODO: Some transcripts overlap, but we can figure out from exon lines which gene to assign a read to.
+			log.Fatalf("The following input SAM record maps to multiple genes in the gtf file:\n%v.", sam.ToString(i))
+		}
+		currGene = getGeneName(overlap[0].(*gtf.Gene))
+		if dna.BasesToString(sc.Bx) != currRow.Bx {
+			if firstTime {
+				firstTime = false
+			} else {
+				printRow(out, currRow, normalizationMap, geneIdSlice, normFlag)
+			}
+			currRow = Row{dna.BasesToString(sc.Bx), make([]float64, len(geneIndex))}
+		}
+		currRow.Counts[geneIndex[currGene]]++ //increment count for gene in currLine
+	}
+	printRow(out, currRow, normalizationMap, geneIdSlice, normFlag) //print the last cell
+	err = out.Close()
+	exception.PanicOnErr(err)
+}
+
+func getGeneName(g *gtf.Gene) string {
+	return g.GeneID
+}
+
+func printRow(out io.Writer, r Row, normalizationMap map[string]float64, geneIdSlice []string, normFlag bool) {
+	if normFlag { //if the user passed in a normalization file, we normalize the count for each gene.
```

Comment:
I considered this but it might be a bit messier as we always want to normalize each row right before we print and we have a couple of calls to the print function (within and after the loop). I can name the function something more descriptive like normAndPrintRow, which might help